### PR TITLE
ElectroHydraulicPTO solver stability

### PR DESCRIPTION
**Pull request by andermi**
Something was causing numerical instability in the ElectroHydraulicPTO "rpm/pressure -> winding current -> torque" solution

To start, I updated the volumetric and mechanical efficiency maps per vendor efficiency maps codified in matlab by @hamilton8415 and brought in a different interpolator: [Splinter](https://github.com/bgrimstad/splinter/blob/master/docs/cpp_interface.md)

@hamilton8415 and I will investigate/resolve issues in this PR

### CODE REVIEW

**Comment by andermi** [comment]
@hamilton8415 something else that was crashing... when the discriminant was negative in the solution for `VBus`

https://github.com/osrf/buoy_sim/blob/af87b95ea6e4e1a3e67385985cec6d9f1adfb8ed/buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicPTO.cpp#L335-L337

**Comment by andermi** [comment]
requires https://github.com/osrf/buoy_msgs/pull/26

**Comment by andermi** [comment]
@hamilton8415 can you check what I did with the new interp and see if the plugin produces the correct output?

**Comment by hamilton8415** [comment]
Will do, I actually wanted to speak with you about how best to incorporate the test comparisons I previously did with the data from the test machine.

**Comment by andermi** [comment]
@mjcarroll @chapulina @quarkytale the new interpolator (if we decide to keep it) has an [MPL-2.0 License](https://github.com/bgrimstad/splinter/blob/master/LICENSE), and I was wondering the best way to credit their project?

**Comment by chapulina** [comment]
IANAL, but reading [Q9](https://www.mozilla.org/en-US/MPL/2.0/FAQ/) here, I think all we need is to include the license file? @tfoote may have a better idea.

**Comment by tfoote** [comment]
Yes, we should bring in the license copy too.

For clarity I would actually suggest packing splinter as a separate package and then just using it in this package to allow easier tracking of potential upstream changes as well as keeping the licensing differentiation clearer.

**Comment by andermi** [comment]
> splinter as a separate package

I've got it in its own package now.
TODO: I assume I should set the license of the package to MPL-2.0 and put the license file at the top level of the package

**Comment by andermi** [comment]
requires https://github.com/osrf/buoy_msgs/pull/28

File: buoy_gazebo/launch/mbari_wec.launch.py [hunk 1]
```diff
@@ -60,7 +65,8 @@ def generate_launch_description():
             pkg_buoy_gazebo,
             'worlds',
             LaunchConfiguration('world_file')
-        ])}.items(),
+        ]),
+            'debugger': LaunchConfiguration('debugger')}.items(),
```
Comment (mjcarroll) [inline buoy_gazebo/launch/mbari_wec.launch.py:69]:
```suggestion
        ]),
        'debugger': LaunchConfiguration('debugger')}.items(),
```

File: buoy_gazebo/launch/mbari_wec_sinusoidal.launch.py [hunk 2]
```diff
@@ -60,7 +65,8 @@ def generate_launch_description():
             pkg_buoy_gazebo,
             'worlds',
             LaunchConfiguration('world_file')
-        ])}.items(),
+        ]),
+            'debugger': LaunchConfiguration('debugger')}.items(),
```
Comment (mjcarroll) [inline buoy_gazebo/launch/mbari_wec_sinusoidal.launch.py:69]:
```suggestion
        ]),
        'debugger': LaunchConfiguration('debugger')}.items(),
```

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp [hunk 3]
```diff
@@ -167,6 +184,101 @@ const std::vector<std::vector<double>> ElectroHydraulicSoln::Neff
...
+    0.9580, 0.9580, 0.9580, 0.9580, 0.9580, 0.9580, 0.9580, 0.9580, 0.9580, 0.9580},
+  {0.1000, 0.7500, 0.8540, 0.8920, 0.9120, 0.9260, 0.9360, 0.9460, 0.9460, 0.9560, 0.9560, 0.9560,
+    0.9560, 0.9560, 0.9560, 0.9560, 0.9560, 0.9560, 0.9560, 0.9560, 0.9560, 0.9560},
+  {0.1000, 0.7400, 0.8460, 0.8880, 0.9080, 0.9240, 0.9340, 0.9440, 0.9440, 0.9540, 0.9540, 0.9540,
+    0.9540, 0.9540, 0.9540, 0.9540, 0.9540, 0.9540, 0.9540, 0.9540, 0.9540, 0.9540},
+  {0.1000, 0.7300, 0.8380, 0.8840, 0.9040, 0.9220, 0.9320, 0.9420, 0.9420, 0.9520, 0.9520, 0.9520,
+    0.9520, 0.9520, 0.9520, 0.9520, 0.9520, 0.9520, 0.9520, 0.9520, 0.9520, 0.9520},
+  {0.1000, 0.7200, 0.8300, 0.8800, 0.9000, 0.9200, 0.9300, 0.9400, 0.9400, 0.9500, 0.9500, 0.9500,
+    0.9500, 0.9500, 0.9500, 0.9500, 0.9500, 0.9500, 0.9500, 0.9500, 0.9500, 0.9500}
+};
```
Comment (mjcarroll) [inline buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:233]: Does it make sense to put these in a format that can be read off disk or are they not changing frequently?

File: buoy_tests/tests/pc_commands_ros_feedback.cpp [hunk 4]
```diff
@@ -201,6 +201,8 @@ TEST_F(BuoyPCTests, PCCommandsInROSFeedback)
     static_cast<int>(node->clock_->now().seconds()),
     static_cast<int>(iterations / 1000.0F));
 
+  std::cout << node->torque_policy_ << std::endl;
```
Comment (mjcarroll) [inline buoy_tests/tests/pc_commands_ros_feedback.cpp:204]: Is this spurious debug output?

File: buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp (hunk 3 above)
Comment (hamilton8415) [reply buoy_gazebo/src/ElectroHydraulicPTO/ElectroHydraulicSoln.hpp:233]: These values are very specific to the hydraulic motor used, so from my point of view it's OK to hard-code them, they should never change for modelling this equipment, if the plugin became more general purpose then maybe, but at this point there's a lot of hardware-specific items in defines and such...

File: buoy_tests/tests/pc_commands_ros_feedback.cpp (hunk 4 above)
Comment (andermi) [reply buoy_tests/tests/pc_commands_ros_feedback.cpp:204]: I can take it out, I just liked seeing in the logs what settings were used

**Comment by andermi** [comment]
I'm gonna say we should merge this in soon as this has been out quite a while and more branches are starting to use it as a base.

File: buoy_tests/tests/pc_commands_ros_feedback.cpp (hunk 4 above)
Comment (andermi) [reply buoy_tests/tests/pc_commands_ros_feedback.cpp:204]: I think I might just leave it in to log policy settings used in the test

Comment (andermi) [reply buoy_tests/tests/pc_commands_ros_feedback.cpp:204]: actually, just discovered the config from param wasn't being loaded... fixing now

**Comment by mjcarroll** [comment]
> We should merge this in soon as this has been out quite a while and more branches are starting to use it as a base.

Agree, what other things need to get in to call this done?

**Comment by andermi** [comment]
> what other things

I think we are good as is. @hamilton8415 has other branches with some other tweaks coming soon

**Comment by quarkytale** [comment]
Works for me, CI should come out green as well

**Review by mjcarroll — APPROVED** [review]

**Comment by andermi** [comment]
@hamilton8415 has requested I cherry-pick new motor efficiency values into this PR before we merge.

**Pull request merged**